lib/modules/manager/gomod/package-tree.ts

```diff
@@ -0,0 +1,187 @@
+  recursivelyGetDependentGoModFiles(packageFileName, graph, visitedModules);
+
+  // deduplicate
+  return Array.from(visitedModules).map(([name, isLeaf]) => ({ name, isLeaf }));
+}
+
+/**
+ * Traverse graph and find dependent go.mod files at any level of ancestry
+ */
+function recursivelyGetDependentGoModFiles(
+  packageFileName: string,
+  graph: Graph,
+  visitedModules: Map<string, boolean>,
+): void {
+  if (visitedModules.has(packageFileName)) {
+    // we have already visited this package file
+    return;
+  }
+
+  const dependents = graph.adjacent(packageFileName);
+
+  if (!dependents || dependents.size === 0) {
+    visitedModules.set(packageFileName, true);
+    return;
+  }
+
+  // Add current node first, then process dependents (pre-order traversal)
+  visitedModules.set(packageFileName, false);
+
+  for (const dependent of dependents) {
+    recursivelyGetDependentGoModFiles(dependent, graph, visitedModules);
+  }
+}
+
+/**
+ * Parse local replace directives from a go.mod file
+ */
+async function parseLocalReplaceDirectives(
+  goModFile: string,
+): Promise<{ targetGoModPath: string | null }[]> {
+  const content = await readLocalFile(goModFile, 'utf8');
+  if (!content) {
+    return [];
+  }
+
+  const directives: { targetGoModPath: string | null }[] = [];
+  const lines = content.split(newlineRegex);
+  const blockRegex = regEx(
+    /^\s*(?<module>[^\s]+)\s*=>\s*(?<replacement>[^\s]+)(?:\s+(?<version>[^\s]+))?(?:\s*\/\/.*)?$/,
+  );
+  let inReplaceBlock = false;
+  for (const line of lines) {
+    const trimmedLine = line.trim();
+
+    // Check if we're starting a multiline replace block
+    if (trimmedLine.startsWith('replace (')) {
+      inReplaceBlock = true;
+      continue;
+    }
+
+    // Check if we're ending a multiline replace block
+    if (inReplaceBlock && trimmedLine === ')') {
+      inReplaceBlock = false;
+      continue;
+    }
+
+    // Parse replace directive (either single line or inside multiline block)
+    let replaceMatches = replaceRegex.exec(line)?.groups;
+
+    // For multiline blocks, the keyword might be missing
+    if (inReplaceBlock && !replaceMatches) {
+      // Try parsing without the replace keyword
+      replaceMatches = blockRegex.exec(line)?.groups;
+    }
+
+    if (replaceMatches) {
+      const { replacement } = replaceMatches;
+      const depName = trimQuotes(replacement);
+
+      // Only process local dependencies (relative paths)
+      if (depName.startsWith('./') || depName.startsWith('../')) {
+        const targetGoModPath = reframeRelativePathToRootOfRepo(
+          goModFile,
+          depName,
+        );
+        const targetGoModFile = upath.join(targetGoModPath, 'go.mod');
+        directives.push({ targetGoModPath: targetGoModFile });
+      }
+    }
+  }
+
+  return directives;
+}
+
+/**
+ * Take the path relative from a go.mod file, and make it relative from the root of the repo
+ */
+function reframeRelativePathToRootOfRepo(
```

Comment:
this function should probably moved to `utils/fs` or `util/tree` and shared between both managers.